src/inmanta/db/versions/v202312190.py

```diff
@@ -35,13 +35,50 @@ async def update(connection: Connection) -> None:
     PRIMARY KEY(environment, resource_id)
 );
 
-INSERT INTO public.resource_persistent_state
- (environment, resource_id, last_deploy, last_success, last_non_deploying_status, last_produced_events)
- SELECT environment, resource_id, last_deploy, last_success, last_non_deploying_status, last_produced_events
- FROM public.resource
- WHERE (environment, model) IN (
-    SELECT environment, max(version) FROM public.configurationmodel WHERE released=true GROUP BY environment
- );
+
+-- The INSERT query uses a subquery to find the maximum model version
```

Comment:
I can't judge the performance of this, @arnaudsjs

Reply:
I ran the SELECT part of the query on mgmt. It executes in about 2.3 seconds. The query plan isn't great but I guess it's just in the nature of the query. I don't really see a way to make it perform better. 2.3s is also not bad for a migration script.

```
                                                                             QUERY PLAN                                                                              
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Gather  (cost=53382.12..117907.04 rows=32 width=154) (actual time=576.725..675.391 rows=4508 loops=1)
   Workers Planned: 2
   Workers Launched: 2
   ->  Hash Join  (cost=52382.12..116903.84 rows=13 width=154) (actual time=579.598..663.079 rows=1503 loops=3)
         Hash Cond: ((r.environment = resource.environment) AND ((r.resource_id)::text = (resource.resource_id)::text) AND (r.model = (max(resource.model))))
         ->  Parallel Seq Scan on resource r  (cost=0.00..63085.62 rows=182362 width=154) (actual time=0.012..57.042 rows=146231 loops=3)
         ->  Hash  (cost=52148.81..52148.81 rows=13332 width=93) (actual time=575.214..575.219 rows=4508 loops=3)
               Buckets: 16384  Batches: 1  Memory Usage: 689kB
               ->  HashAggregate  (cost=51882.17..52015.49 rows=13332 width=93) (actual time=572.881..573.859 rows=4508 loops=3)
                     Group Key: resource.environment, resource.resource_id
                     Batches: 1  Memory Usage: 1425kB
                     Worker 0:  Batches: 1  Memory Usage: 1425kB
                     Worker 1:  Batches: 1  Memory Usage: 1425kB
                     ->  Nested Loop  (cost=0.55..48599.66 rows=437668 width=93) (actual time=0.046..461.675 rows=438694 loops=3)
                           ->  Seq Scan on configurationmodel  (cost=0.00..34.11 rows=211 width=20) (actual time=0.013..0.512 rows=207 loops=3)
                                 Filter: released
                           ->  Index Only Scan using resource_pkey on resource  (cost=0.55..223.02 rows=715 width=93) (actual time=0.016..2.055 rows=2119 loops=621)
                                 Index Cond: ((environment = configurationmodel.environment) AND (model = configurationmodel.version))
                                 Heap Fetches: 394218
 Planning Time: 1.729 ms
 Execution Time: 676.022 ms
```